### ISSUE TITLE
Video wallpaper - how to enable it?

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [Alaory/WallMe-Wallpaper#31]
I've chosen some video wallpapers and I've set the app as a live wallpaper, but I still see black content.
How can I make it show videos?
And is there any way to have horizontal scrollable wallpapers (image&video) ? I don't see the option for this either...

**Comment by Alaory** [comment]
Can you give me info about your device?

**Comment by Alaory** [comment]
Can you try loading a local video wallpaper? 
Favorite -> add wallpaper -> select your wallpaper. 
Try that and tell me the results

**Comment by AndroidDeveloperLB** [comment]
What do you mean "not planned"?
It says it supports it, no?
Please show how to do it. Maybe I'm using it wrong

**Comment by Alaory** [comment]
Can you explain your problem in depth? like what is the source of your video wallpaper e.g (from your device or reddit) 
what is the video format?

**Comment by AndroidDeveloperLB** [comment]
It was from reddit. Can you please demonstrate how to do both?
The UX is confusing to me, so maybe I did it wrong.

**Comment by Alaory** [comment]
If you did all of that and the wallpaper still look black. is it possible for you to provide me the cat log of the app?

**Comment by Alaory** [comment]
When you click on a video wallpaper before you set it as one. Does the video play?

**Comment by AndroidDeveloperLB** [comment]
The preview plays fine. It's the live wallpaper itself that doesn't show anything.

Here, a video of what I did:
[scrcpy_2023-03-15_12-15-34.zip](https://github.com/Alaory/WallMe-Wallpaper/files/10978538/scrcpy_2023-03-15_12-15-34.zip)


Please explain what I did wrong. I got it from F-Droid:
https://f-droid.org/packages/com.alaory.wallmewallpaper/

**Comment by Alaory** [comment]
Haha ok got it might be dumb and it is a an issue that i don't know how to fix it on the pixel yet. but when you preview the video did you try to pull up the floating navigation. the one under your system navigation?

**Comment by AndroidDeveloperLB** [comment]
Oh I didn't notice it. It's behind the system navigation.
It's not related to Pixel 6. Here, it happens on a different device too (Honor magic 4 pro with Android 12) : 

![image](https://user-images.githubusercontent.com/5357526/225282779-b8fde6b7-d818-4427-a62f-44e9cc1858dc.png)

What does the "heart" icon do? It doesn't seem to do anything except adding to the list.
The left button seems to apply the video.
So, this live wallpaper supports images, GIF/WEBP animations and video files?
Does it also have horizontal scrolling feature (in most cases the aspect ratio isn't exactly the same as the device, after all) ?

**Comment by Alaory** [comment]
Huh guess gotta fix the navigation soon. the heart icon only adds/removes the wallpaper to/from the favorite list. The aspect ratio will be the same as when you set it. there are a lot of problems with the app and i can't work on all of them at the same time. 🫠

**Comment by AndroidDeveloperLB** [comment]
The aspect ratio is actually bad for MP4 file I chose from the device, in both the preview and on the live wallpaper. It doesn't keep shot the video properly at all. I think it doesn't crop at all. Just stretches it.

Please consider these changes:

1. Make the bottom UI more visible. I really didn't notice it at all. I thought there is something weird here as I looked at the screenshots and I thought that maybe I got a different version.
2. Add toggle of horizontal scrolling, for all kinds of media. Most launchers support it. You can add a warning that some launchers don't (I think of Samsung, and that's intentional for some reason).
3. Add a new option to choose image/video from the file picker. I don't see it. Only "settings", "favorites", "reddit" and "filter". The way to reach it currently is actually via "favorites", and instead of adding to favorites I choose to apply it.
4. Fix the bug of showing a video. 

I've checked the code a bit and found this:

https://github.com/Alaory/WallMe-Wallpaper/blob/master/app/src/main/java/com/alaory/wallmewallpaper/wallpaper/livewallpaper.kt

I thought it's impossible to switch between engines. How do you do it?
Also, how come "imageengine" (shouldn't it be "ImageEngine"?) is defined there, but isn't used in this class?

**Comment by Alaory** [comment]
I still haven't enable the "ImageEngine" ignore the fancy name i plan on using it  with the wallpaper_changer so it could show live wallpapers.

**Comment by AndroidDeveloperLB** [comment]
What do you mean "so it could show live wallpapers" ? It looks like it's to show images on the live wallpaper, no?
I'm mostly curious about how you switched between the engines. I thought this is impossible. There is no API for this.
No?

**Comment by Alaory** [comment]
I just kill the old engine and before the android system request the engine to start again i replace it with a new one ("based on the wallpaper") . Android system doesn't request the engine directly it asks the WallpaperService for the engine when that happens the above happens :)

**Comment by AndroidDeveloperLB** [comment]
How do you kill it? You use a different process for the live wallpaper and somehow trigger killing it?
Or something else?

Wouldn't it be possible and better to avoid killing and use the same engine?
I know that Muzei somehow shows images without using Canvas, for example.

**Comment by Alaory** [comment]
https://github.com/Alaory/WallMe-Wallpaper/blob/master/app/src/main/java/com/alaory/wallmewallpaper/Image_Activity.kt#L344

**Comment by AndroidDeveloperLB** [comment]
Wait, but the docs say it will revert to the default.
Instead, it stays on the current live wallpaper, and just triggers the re-creation of the engine?

**Comment by Alaory** [comment]
No it does revert to default wallpaper :)

**Comment by AndroidDeveloperLB** [comment]
I thought you've found a workaround. 
So how does the "wallpaper changer" work in case you have a mixed list of videos and images?
It will skip in case it needs to switch engine?

**Comment by AndroidDeveloperLB** [comment]
Looking at the code and UI, I think the "wallpaper changer" isn't for the live wallpaper. Only for normal ones (images).
So the live wallpaper sadly always asks to be applied via the OS confirmation on each media that I choose to show on it...

**Comment by Alaory** [comment]
Gonna find a way somehow  :)

**Comment by AndroidDeveloperLB** [comment]
I wonder how you will do this. 
Sadly I'm not an expert on OpenGL (I had some knowledge on the first version , but it has changed completely on later versions) and on live wallpapers (I know more about using the Canvas).
Even for LibGdx, it was a very long time ago since I did anything with it, and I have no idea how to use it in a live wallpaper.

I guess using LibGdx is going too far, but if you know how to use it, I think it could help you because it should support videos and images already.

**Comment by AndroidDeveloperLB** [comment]
What if you check how Muzei works?

**Comment by Alaory** [comment]
I did have a look at it. I might do it the way muzei did it

**Comment by AndroidDeveloperLB** [comment]
@Alaory They use it for images alone. And you have for video. So maybe possible to merge.
They also have horizontal scrolling for images, so maybe it will work for video too.
There is another repository that you could check of this:
https://github.com/AlynxZhou/alynx-live-wallpaper

It's for video and it has horizontal scrolling too, but as I've noticed, the horizontal scrolling might not look well in some cases:
https://github.com/AlynxZhou/alynx-live-wallpaper/issues/21

**Comment by AndroidDeveloperLB** [comment]
I also checked Muzei, but it seems so complicated. Google tends to complicate things a lot, even for samples...

**Comment by Alaory** [comment]
I think there are forks of muzei that support video wallpaper

**Comment by AndroidDeveloperLB** [comment]
Perhaps OpenGL is more efficient:
https://stackoverflow.com/a/6486679/878126

Muzei is so complex to deal with, though.
Can you please consider making a new repository that has the bare-bone of video&image handling from Muzei, and yet without all the extra stuff they have there (images sources providers, blurring, ...) , including the horizontal scrolling? 
I can donate you if you will do this as simple as possible (but working well), if you wish.

**Comment by Alaory** [comment]
If i'm gonna use openGl might as well use cpp or rust since i more comfortable with them. But for now i'll try find a way to use Kotlin to handle the live wallpapers. As for muzei i'll think of that as a future project since it would take a long time for me to read the Muzei repo.
Thanks again for all the help you provided me

**Comment by Alaory** [comment]
Even worse I have to compile librarys for each architecture. Thanks for info i though JNI was faster didn't know ART was better than DALVIK.

**Comment by AndroidDeveloperLB** [comment]
I assume it depends on the case. Maybe in extreme cases you'd prefer C/C++. 
But in most cases Java/Kotlin should be fine. I've seen some benchmarks and they seem fine.
I'm pretty sure that real game engines have plenty of optimizations, so they use C/C++ behind the hood, even though the developers that use them probably don't use C/C++.

**Comment by Alaory** [comment]
I don't think that muzei does use c++. 
If jvm is better than c++ then i would use it. Now i am thinking of using libgdx. But I think it is an overuse of using libgdx just to render video or gif

**Comment by AndroidDeveloperLB** [comment]
I'm sure there are still cases that C/C++ is better. 
But for most cases, Java/Kotlin should be fine. 
About LibGdx , sure it's meant for games, but it can be used for this purpose too. 
And maybe you could extend the functionality to more than images/animation/videos some day. Not sure what, but you could. 

If you want, I can donate you because I tried in the past to understand how to do it, and failed . Please, though, create a new repository with the bare minimum, as POC, before you add this change to the current one. I want to learn from a simple example, as simple as possible... 
:)

**Comment by Alaory** [comment]
I still have to understand how to use libgdx i had a look at their docs it didn't seem that hard. For now i'll mainly focus on the app bugs i still need to fix them

**Comment by AndroidDeveloperLB** [comment]
Wow you already started!
Thank you!

**Comment by AndroidDeveloperLB** [comment]
I've decided to give it a try too, and wrote about this here, in case anyone who knows LibGdx could help with video on a live wallpaper:
https://www.reddit.com/r/libgdx/comments/11uo9zv/help_showing_video_and_image_on_a_live_wallpaper/

I got it to show a center-cropped image, but I failed to make it work properly for horizontal scrolling. 
Probably something wrong with my calculations as it worked fine on Canvas, but here things are probably different.

Here's the project, if you wish to try it out:

[Time Flies Live Wallpaper.zip](https://github.com/Alaory/WallMe-Wallpaper/files/11009411/Time.Flies.Live.Wallpaper.zip)

I've noticed another thing:
When I use the profiler, it shows that it constantly uses the CPU. Very low percentage (3%), but still...
When using the Canvas method, and when I've launched Muzei, it uses the CPU only when scrolling.
LibGdx is usually for games so it makes sense, but maybe it can be optimized to render only when needed, as I've seen here:
https://libgdx.com/wiki/graphics/continuous-and-non-continuous-rendering
